### PR TITLE
Limit the number of displayed changes (bsc#1044777)

### PR DESCRIPTION
**Pull request by lslezak**
- Rendering a huge change log might cause a freeze for long time, see https://bugzilla.suse.com/show_bug.cgi?id=1044777.
- With this fix displaying the the *kernel-default* changes dropped from ~14 seconds to ~1 second.
- There is a note about missing items at the end, see the screenshot below
- 2.45.13

screenshot_changelog

# Additional Tasks

- [x] Docker support for the SP3 branch to make Travis pass

### CODE REVIEW

File: package/libyui-qt-pkg.changes [hunk 1]
```diff
@@ -1,4 +1,11 @@
 -------------------------------------------------------------------
+Fri Jun 23 12:05:27 UTC 2017 - lslezak@suse.cz
+
+- Limit the number of displayed changes, rendering a huge change
```
Comment (kobliha) [inline package/libyui-qt-pkg.changes]: Maybe some number here .... :)?

File: src/YQPkgChangeLogView.cc [hunk 2]
```diff
@@ -110,10 +128,24 @@ QString YQPkgChangeLogView::changeLogTable( const zypp::Changelog & changeLog )
...
 
     return html.isEmpty() ? "" : table( html );
 }
 
+QString YQPkgChangeLogView::notDisplayedChanges(int missing, const std::string &pkg)
+{
+    // TRANSLATORS: The package change log is too long to display, only the latest
+    // changes are displayed. %1 is the number of the items which are not displayed,
+    // %2 contains a command for getting the full changes manually.
+    QString msg = _("(Next %1 change entries are not displayed. Run command \""
```
Comment (mvidner) [inline src/YQPkgChangeLogView.cc]: `"%1 more change entries... . Run \"%2\" to see...`

File: src/YQPkgChangeLogView.cc [hunk 3]
```diff
@@ -48,7 +48,9 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "YQi18n.h"
 #include "utf8.h"
 
-
+// how many change log entries should be displayed at most,
+// displaying huge changes takes too much time (bsc#1044777)
+static const int MAX_DISPLAYED_CHANGES = 512;
```
Comment (jreidinger) [inline src/YQPkgChangeLogView.cc]: I still think that 512 is increadible number, it is not number of lines, it is number of changes, so in reality around 2048 lines, which is too much from my POV. I suggest to limit it to 16 or 64. It is reasonable number of the latest changes for me, that still make sense to view in packager.

File: src/YQPkgChangeLogView.cc [hunk 4]
```diff
@@ -73,30 +75,46 @@ YQPkgChangeLogView::showDetails( ZyppSel selectable )
...
+
     QString html = htmlStart();
     html += htmlHeading( selectable, false );
 
     ZyppPkg installed = tryCastToZyppPkg( selectable->installedObj() );
 
     if ( installed )
     {
-	html += changeLogTable( installed->changelog() );
+	    html += changeLogTable( installed->changelog() );
```
Comment (mvidner) [inline src/YQPkgChangeLogView.cc]: inconsistent indentation

File: src/YQPkgChangeLogView.h [hunk 5]
```diff
@@ -81,6 +81,14 @@ class YQPkgChangeLogView : public YQPkgGenericDetailsView
...
      **/
     QString changeLogTable( const zypp::Changelog & changeLog ) const;
+
+    /**
+     * Format an info message about not displayed changes.
+     * @param  missing Number of missing entris
+     * @param  pkg     Package name (with version)
+     * @return         Translated message
+     */
+    QString notDisplayedChanges(int missing, const std::string &pkg);
```
Comment (jreidinger) [inline src/YQPkgChangeLogView.h]: ABI incompatible change ( should increase so minor version ).

File: src/YQPkgChangeLogView.cc (hunk 3 above)
Comment (lslezak) [inline src/YQPkgChangeLogView.cc]: Well, I wanted to keep the current behavior as much as possible. I.e. display the complete change log if possible, limit the length only in the pathological cases when it is really too much.

So the question is how much we want to be backward compatible, some people might consider this as a regression...

Comment (lslezak) [inline src/YQPkgChangeLogView.cc]: BTW for `kernel-default` that 512 entries are just until October 2016, that's roughly 8 months ago, not too much....

File: src/YQPkgChangeLogView.h (hunk 5 above)
Comment (lslezak) [inline src/YQPkgChangeLogView.h]: Um, does it make sense here actually? 

- the library is `dlopen`ed and the API used by libyui is not changed
- the library is not used by any other consumer

Moreover see [this comment](https://github.com/libyui/libyui-qt-pkg/blob/master/VERSION.cmake#L6-L7).

@mvidner how should be this handled? Is there any change requited?

File: package/libyui-qt-pkg.changes (hunk 1 above)
Comment (lslezak) [inline package/libyui-qt-pkg.changes]: Ok, lets discuss the number first :wink: , see below...

File: src/YQPkgChangeLogView.h (hunk 5 above)
Comment (mvidner) [inline src/YQPkgChangeLogView.h]: @jreidinger IIUC this is a *compatible* ABI change

There are two ABIs here: user - libyui and libyui - plugins. The current versioning scheme is not really adequate to express it, and I don't know how to fix it.

File: src/YQPkgChangeLogView.cc (hunk 3 above)
Comment (jsrain) [inline src/YQPkgChangeLogView.cc]: Let's try suggested 512 entries, if there are any complaints, we can double the number. After all, the main use case for changelog is IMO "waht's new compared to my previously installed version" for applying maintenance updates.

Comment (kobliha) [inline src/YQPkgChangeLogView.cc]: Let's use 512 PLS. Most of packages will have fewer entries anyway. For the rest, the speed is already good enough.

**Pull request merged**